fix(platform-server): destroy `PlatformRef` when error happens during the `bootstrap()` phase, e.g. in `APP_INIIALIZER` - to prevent memory leaks in SSR

Platonn opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [N/A] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
`PlatformRef` is not destroyed if an error happens during the `bootstrap()` phase (e.g. in `APP_INITIALIZER`). Then the main app's injector is not destroyed and therefore `ngOnDestroy` hooks of singleton services were not called on the end (failure) of SSR. 

Note: This could lead to possible memory leaks in custom SSR apps, if their singleton services' `ngOnDestroy` hooks contained an important teardown logic (e.g. unsubscribing from RxJS observable).

Issue Number: https://github.com/angular/angular/issues/58111


## What is the new behavior?
- `PlatformRef` is guaranteed to be destroyed in `renderApplication()` and `renderModule()` functions of `@angular/platform-server` both in the case of successful render and in the case of a failure (resulting in a rejected promise)
- `PLATFORM_DESTROY_LISTENERS` that destroys the app's main injector is now set not only in case of `renderApplication()`, but also in case of `renderModule()`


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche commented:
Hi Krzysztof, thanks for looking into this. Could you please rebase your PR on the `main` branch (so we can run the CI)

Platonn commented:
Thank you @JeanMeche for your reply.
I've rebased with `main` just now and resolved the code conflicts.

JeanMeche commented on packages/platform-server/test/integration_spec.ts:
```diff
@@ -1076,6 +1079,153 @@ class HiddenModule {}
+              createMyServerApp(true),
+              FailingAppInitializerProviders,
+            );
+            debugger;
```
You forgot this one

Platonn replied:
🤦‍♂️ Thank you for noticing it. Now removed.

Platonn committed:
fix(platform-server): destroy `PlatformRef` when error happens during the `bootstrap()` phase

The `bootstrap()` phase might fail e.g. due to an rejected promise in some `APP_INIIALIZER`.
If `PlatformRef` is not destroyed, then the main app's injector is not destroyed and therefore `ngOnDestroy` hooks of singleton services is not called on the end (failure) of SSR.

This could lead to possible memory leaks in custom SSR apps, if their singleton services' `ngOnDestroy` hooks contained an important teardown logic (e.g. unsubscribing from RxJS observable).

Note: I needed to fix by the way another thing too: now we destroy `moduleRef` when `platformInjector` is destroyed - by setting a `PLATFORM_DESTROY_LISTENER`

fixes #58111

JeanMeche commented:
The changes are breaking the `elements` pacakge, can you have a look ?

Platonn commented:
Thank you @JeanMeche for letting me know.
Now pushed a fix.

Explanation:
- Before this PR
  - the `isApplicationBootstrapConfig()` was duck-typing based on the presence of the property `platformInjector`.
- In this PR initially
  -  I've changed the logic to check the property `rootComponent` instead, because I wanted to pass the `platformInjector` in all cases - also in the NgModule flow (in order to setup the `PLATFORM_DESTROY_LISTENER` later in the `bootstrap()` function).
   - It turned out to break the behavior of the public API function `createApplication()` which was tested inside the `elements` package. The test failed, because no `rootComponent` was passed in that test case setup as a param of `createApplication`. Then `isApplicationBootstrapConfig()` returned `false` which let into executing NgModule-flow of `bootstrap()`. Despite someone could perceive omitting the `rootComponent` as an edge case usage of `createApplication()`, I broke that default behavior. So needed to fix it
- The fix
  - So now I've fixed the problem by no-longer duck-typing inside the `isApplicationBootstrapConfig()` function. Instead, I started passing explicitly a newly-introduced config property `BootstrapKind` as a required param of the `bootstrap()` function. btw. It's not a breaking change, because the signature of the `bootstrap()` function is not a part of the public API, as far as I can understand.

JeanMeche commented on packages/core/src/platform/bootstrap.ts:
```diff
@@ -26,21 +26,32 @@ import {Injector} from '../di';
 import {stringify} from '../util/stringify';
 
-export interface ModuleBootstrapConfig<M> {
+export enum BootstrapKind {
```
nit:
fwiw, if the enum is not exported outside of the package, we can make it a `const enum` with a numeric value. (It'll save a few bytes)

Platonn replied:
Thank you for a hint.
Pushed the fix now.

alan-agius4 commented on packages/platform-server/src/utils.ts:
```diff
@@ -210,18 +210,20 @@ async function _render(platformRef: PlatformRef, applicationRef: ApplicationRef)
 
   appendServerContextInfo(applicationRef);
-  const output = platformState.renderToString();
+  return platformState.renderToString();
```
NIT

```suggestion
  
  return platformState.renderToString();
```

alan-agius4 commented on packages/platform-server/src/utils.ts:
```diff
@@ -210,18 +210,20 @@ async function _render(platformRef: PlatformRef, applicationRef: ApplicationRef)
+ * Destroy the application in a macrotask, this allows pending promises to be settled and errors
+ * to be surfaced to the users.
+ */
+async function asyncDestroyPlatform(platformRef: PlatformRef): Promise<void> {
   await new Promise<void>((resolve) => {
```
Avoid creating an extra microtask.

```suggestion
function asyncDestroyPlatform(platformRef: PlatformRef): Promise<void> {
  return new Promise<void>((resolve) => {
```

Platonn replied:
Thank you. done

alan-agius4 reviewed (approved):
LGTM, just a couple of NITs

Platonn commented on packages/core/src/platform/bootstrap.ts:
```diff
@@ -26,21 +26,32 @@ import {Injector} from '../di';
   config: ApplicationBootstrapConfig | ModuleBootstrapConfig<unknown>,
 ): config is ApplicationBootstrapConfig {
-  return !!(config as ApplicationBootstrapConfig).platformInjector;
+  return config.bootstrapKind === BootstrapKind.Application;
```
Thank you. done

AndrewKushnir commented:
@Platonn thanks for creating this PR 👍 It looks like there is a failing test (see [the `test` CI job output](https://github.com/angular/angular/actions/runs/11237605771/job/31241076382?pr=58112)). Could you please take a look at it when you get a chance?

JeanMeche commented:
@AndrewKushnir It looks like the failing test is flaky.

AndrewKushnir commented:
[Presubmit](http://test/OCL:683881745:BASE:683881771:1728448838691:abd9cd7b).

AndrewKushnir commented:
**Caretaker note**: presubmit looks good (only pre-existing failing targets), this PR is ready for merge.

devversion commented:
This PR was merged into the repository by commit 9e82559de4e99a1aedf645a05b01fc08d3f4b1b1.

The changes were merged into the following branches: main

devversion commented:
@Platonn @JeanMeche this seems to have conflicts with the patch branch, so I only merged for v19. Please send another PR if necessary— I tried creating a patch port but I don't know how conflicts should be solved

JeanMeche commented:
@devversion Patch #58135
